chain/gap/fill.go

```diff
@@ -0,0 +1,101 @@
+package gap
+
+import (
+	"context"
+
+	"github.com/filecoin-project/sentinel-visor/chain"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/model/visor"
+	"github.com/filecoin-project/sentinel-visor/storage"
+	"github.com/go-pg/pg/v10"
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+)
+
+var log = logging.Logger("visor/gap")
+
+type GapFiller struct {
+	DB                   *storage.Database
+	opener               lens.APIOpener
+	name                 string
+	minHeight, maxHeight uint64
+	tasks                []string
+}
+
+func NewGapFiller(o lens.APIOpener, db *storage.Database, name string, maxHeight, minHeight uint64, tasks []string) *GapFiller {
+	return &GapFiller{
+		DB:        db,
+		opener:    o,
+		name:      name,
+		maxHeight: maxHeight,
+		minHeight: minHeight,
+		tasks:     tasks,
+	}
+}
+
+func (g *GapFiller) Run(ctx context.Context) error {
+	gaps, err := g.queryGaps(ctx)
+	if err != nil {
+		return err
+	}
+	fillLog := log.With("type", "fill")
+	fillLog.Infow("run", "count", len(gaps))
+
+	for idx, gap := range gaps {
+		// TODO we could optimize here by collecting all gaps at same height and launching a single instance of the walker for them
+		indexer, err := chain.NewTipSetIndexer(g.opener, g.DB, 0, g.name, []string{gap.Task})
+		if err != nil {
+			gap.Status = err.Error()
+			log.Errorw("fill failed", "height", gap.Height, "error", err.Error())
+		} else {
+			walker := chain.NewWalker(indexer, g.opener, gap.Height, gap.Height)
```

Comment:
Is this only ever going to deal with single height gaps? What if we are missing 10 epochs in a row?

Also, many tasks require 2 tipsets to operate to perform the diff (we could simply fetch parent if we haven't walked it but currently we don't do that)

Reply:
> Is this only ever going to deal with single height gaps? What if we are missing 10 epochs in a row?

I have a TODO above that calls this out. An enhancement to this logic would find the longest contiguous gap and walk that. I haven't written it yet as the code works fine without the optimization.

> Also, many tasks require 2 tipsets to operate to perform the diff (we could simply fetch parent if we haven't walked it but currently we don't do that)

I think this happens implicitly via the WalkChain method, but maybe I am misunderstanding the point you are making:
First call fails to find a parent and set `lastTipSet`
https://github.com/filecoin-project/sentinel-visor/blob/ba4c266e345c7f0018bd609c566d19324874e587/chain/walker.go#L75-L78

Second call has a parent and:
https://github.com/filecoin-project/sentinel-visor/blob/ba4c266e345c7f0018bd609c566d19324874e587/chain/walker.go#L93-L96

Are you stating the indexer could fetch the parent if it's null?

Reply:
Yes if parent  == nil we could fetch it quite easily and cheaply.

Reply:
Addressed via: https://github.com/filecoin-project/sentinel-visor/pull/653